File: web-tunnel/client-android/app/src/main/java/ai/webtunnel/mobile/ClientConfig.kt
package ai.webtunnel.mobile

import android.util.Base64
import org.json.JSONObject
import java.security.MessageDigest
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import java.util.UUID

private const val CLIENT_CONFIG_PREFIX = "wtc1:"
private const val CONFIG_KEY_LABEL = "web-tunnel encrypted client config v1"
private const val NONCE_BYTES = 12

data class ClientConfigServerPeer(
  val chatId: Long,
  val chatType: String,
  val label: String,
)

data class ClientConfigPayload(
  val clientId: String,
  val clientName: String,
  val createdAt: Long,
  val defaultSocksPort: Int,
  val baleSession: BaleSession,
  val serverPeer: ClientConfigServerPeer,
  val serverUuid: String,
  val serverFingerprint: String?,
)

fun decodeClientConfig(input: String): ClientConfigPayload {
  val trimmed = input.trim()
  require(trimmed.startsWith(CLIENT_CONFIG_PREFIX)) { "client config must start with wtc1:" }
  val encoded = trimmed.removePrefix(CLIENT_CONFIG_PREFIX)
  val packed = Base64.decode(encoded, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
  require(packed.size > NONCE_BYTES) { "client config payload is too short" }
  val nonce = packed.copyOfRange(0, NONCE_BYTES)
  val encrypted = packed.copyOfRange(NONCE_BYTES, packed.size)
  val key = MessageDigest.getInstance("SHA-256").digest(CONFIG_KEY_LABEL.toByteArray(Charsets.UTF_8))
  val cipher = Cipher.getInstance("AES/GCM/NoPadding")
  cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
  val plain = cipher.doFinal(encrypted).toString(Charsets.UTF_8)
  val raw = JSONObject(plain)
  require(raw.optInt("schema") == 1) { "unsupported client config schema" }
  val sessionRaw = raw.getJSONObject("baleSession")
  val serverRaw = raw.getJSONObject("serverPeer")
  val carrier = raw.optString("carrier")
  val serverUuid = raw.getString("serverUuid")
  require(carrier == "webrtc") { "client config carrier must be webrtc" }
  require(isValidUuid(serverUuid)) { "client config has invalid server uuid" }
  return ClientConfigPayload(
    clientId = raw.getString("clientId"),
    clientName = raw.getString("clientName"),
    createdAt = raw.optLong("createdAt", System.currentTimeMillis()),
    defaultSocksPort = raw.optInt("defaultSocksPort", 1080),
    baleSession = BaleSession(
      jwt = sessionRaw.getString("jwt"),
      userId = sessionRaw.getString("userId").toLong(),
      userName = sessionRaw.optString("userName").ifBlank { null },
      userAccessHash = sessionRaw.getString("userAccessHash").toLong(),
    ),
    serverPeer = ClientConfigServerPeer(
      chatId = serverRaw.getLong("chatId"),
      chatType = serverRaw.getString("chatType"),
      label = serverRaw.optString("label").ifBlank { "server" },
    ),
    serverUuid = serverUuid,
    serverFingerprint = raw.optString("serverFingerprint").ifBlank { null },
  )
}

private fun isValidUuid(value: String): Boolean {
  return try {
    UUID.fromString(value).toString().equals(value, ignoreCase = true)
  } catch (_: Throwable) {
    false
  }
}

File: web-tunnel/shared/src/client-config.ts
export const CLIENT_CONFIG_PREFIX = 'wtc1:';

const enc = new TextEncoder();
const dec = new TextDecoder();
const CONFIG_KEY_LABEL = 'web-tunnel encrypted client config v1';
const NONCE_BYTES = 12;

export interface ClientConfigBaleSession {
  jwt: string;
  userId: string;
  userName: string | null;
  userAccessHash: string;
}

export interface ClientConfigServerPeer {
  chatId: number;
  chatType: 'PRIVATE' | 'GROUP';
  label: string;
}

export interface WebTunnelClientConfigV1 {
  schema: 1;
  clientId: string;
  clientName: string;
  createdAt: number;
  carrier: 'webrtc';
  defaultSocksPort: number;
  baleSession: ClientConfigBaleSession;
  serverPeer: ClientConfigServerPeer;
  serverUuid: string;
  serverFingerprint: string | null;
}

export async function encodeClientConfig(payload: WebTunnelClientConfigV1): Promise<string> {
  validatePayload(payload);
  const nonce = crypto.getRandomValues(new Uint8Array(NONCE_BYTES));
  const plain = enc.encode(JSON.stringify(payload));
  const key = await configKey();
  const encrypted = new Uint8Array(await crypto.subtle.encrypt({ name: 'AES-GCM', iv: nonce }, key, plain));
  const packed = new Uint8Array(nonce.byteLength + encrypted.byteLength);
  packed.set(nonce, 0);
  packed.set(encrypted, nonce.byteLength);
  return CLIENT_CONFIG_PREFIX + base64UrlEncode(packed);
}

export async function decodeClientConfig(input: string): Promise<WebTunnelClientConfigV1> {
  const trimmed = input.trim();
  if (!trimmed.startsWith(CLIENT_CONFIG_PREFIX)) {
    throw new Error('client config must start with wtc1:');
  }
  const packed = base64UrlDecode(trimmed.slice(CLIENT_CONFIG_PREFIX.length));
  if (packed.byteLength <= NONCE_BYTES) throw new Error('client config payload is too short');
  const nonce = packed.slice(0, NONCE_BYTES);
  const encrypted = packed.slice(NONCE_BYTES);
  const key = await configKey();
  const plain = new Uint8Array(await crypto.subtle.decrypt({ name: 'AES-GCM', iv: nonce }, key, encrypted));
  const parsed = JSON.parse(dec.decode(plain)) as WebTunnelClientConfigV1;
  validatePayload(parsed);
  return parsed;
}

async function configKey(): Promise<any> {
  const digest = await crypto.subtle.digest('SHA-256', enc.encode(CONFIG_KEY_LABEL));
  return crypto.subtle.importKey('raw', digest, 'AES-GCM', false, ['encrypt', 'decrypt']);
}

function validatePayload(payload: WebTunnelClientConfigV1): void {
  if (!payload || payload.schema !== 1) throw new Error('unsupported client config schema');
  if (!payload.clientId || typeof payload.clientId !== 'string') throw new Error('client config missing client id');
  if (!payload.clientName || typeof payload.clientName !== 'string') throw new Error('client config missing client name');
  if (payload.carrier !== 'webrtc') throw new Error('client config carrier must be webrtc');
  if (!payload.baleSession || typeof payload.baleSession.jwt !== 'string' || !payload.baleSession.jwt) {
    throw new Error('client config missing Bale session');
  }
  if (!payload.baleSession.userId || !payload.baleSession.userAccessHash) {
    throw new Error('client config missing Bale account identity');
  }
  if (!payload.serverPeer || typeof payload.serverPeer.chatId !== 'number' || !payload.serverPeer.chatType) {
    throw new Error('client config missing server peer');
  }
  if (!payload.serverUuid || typeof payload.serverUuid !== 'string') throw new Error('client config missing server uuid');
  if (!isValidUuid(payload.serverUuid)) throw new Error('client config has invalid server uuid');
  if (payload.serverPeer.chatType !== 'PRIVATE' && payload.serverPeer.chatType !== 'GROUP') {
    throw new Error('client config has unsupported server peer type');
  }
}

function isValidUuid(value: string): boolean {
  return /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(value);
}

function base64UrlEncode(bytes: Uint8Array): string {
  return Buffer.from(bytes)
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/g, '');
}

function base64UrlDecode(value: string): Uint8Array {
  const padded = value.replace(/-/g, '+').replace(/_/g, '/')
    + '='.repeat((4 - (value.length % 4)) % 4);
  return new Uint8Array(Buffer.from(padded, 'base64'));
}
